File: catalyst-common/src/main/java/org/anvilpowered/catalyst/common/command/ToggleProxyChat.java
/*
 * Catalyst - AnvilPowered
 *   Copyright (C) 2020
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see https://www.gnu.org/licenses/.
 */

package org.anvilpowered.catalyst.common.command;

import com.google.inject.Inject;
import com.mojang.brigadier.context.CommandContext;
import org.anvilpowered.anvil.api.registry.Registry;
import org.anvilpowered.anvil.api.util.PermissionService;
import org.anvilpowered.anvil.api.util.TextService;
import org.anvilpowered.catalyst.api.plugin.PluginMessages;
import org.anvilpowered.catalyst.api.registry.CatalystKeys;
import org.anvilpowered.catalyst.api.service.ChatService;

public class ToggleProxyChat<
    TString,
    TPlayer extends TCommandSource,
    TCommandSource> {

    @Inject
    private ChatService<TString, TPlayer, TCommandSource> chatService;

    @Inject
    private PermissionService permissionService;

    @Inject
    private PluginMessages<TString> pluginMessages;

    @Inject
    private Registry registry;

    @Inject
    private TextService<TString, TCommandSource> textService;

    public int execute(CommandContext<TCommandSource> context, Class<?> playerClass) {
        if (!permissionService.hasPermission(context.getSource(),
            registry.getOrDefault(CatalystKeys.TOGGLE_CHAT_PERMISSION))) {
            textService.send(pluginMessages.getNoPermission(), context.getSource());
            return 0;
        }
        if (!playerClass.isAssignableFrom(context.getSource().getClass())) {
            textService.send(textService.of("Player only command!"), context.getSource());
            return 0;
        }
        TPlayer player = (TPlayer) context.getSource();

        if (chatService.isDisabledForUser(player)) {
            chatService.toggleChatForUser(player);
            sendToggle(context.getSource(), true);
            return 1;
        }
        chatService.toggleChatForUser(player);
        sendToggle(context.getSource(), false);
        return 1;
    }

    private void sendToggle(TCommandSource source, boolean enabled) {
        String setting = enabled ? "enabled" : "disabled";
        textService.builder()
            .appendPrefix()
            .green().append("Proxy-Wide chat has been ")
            .gold().append(setting)
            .sendTo(source);
        textService.builder()
            .yellow().append("Note: This only effects YOU")
            .sendTo(source);
    }
}

File: catalyst-common/src/main/kotlin/org/anvilpowered/catalyst/common/command/ToggleProxyChat.kt
/*
 *   Catalyst - AnvilPowered
 *   Copyright (C) 2021
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see https://www.gnu.org/licenses/.
 */
package org.anvilpowered.catalyst.common.command

import com.google.inject.Inject
import com.mojang.brigadier.context.CommandContext
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import org.anvilpowered.anvil.api.misc.sendTo
import org.anvilpowered.anvil.api.plugin.PluginInfo
import org.anvilpowered.anvil.api.util.PermissionService
import org.anvilpowered.catalyst.api.plugin.PluginMessages
import org.anvilpowered.catalyst.api.registry.CatalystKeys
import org.anvilpowered.catalyst.api.service.ChatService
import org.anvilpowered.registry.Registry

class ToggleProxyChat<TPlayer : TCommandSource, TCommandSource> @Inject constructor(
    private val chatService: ChatService<TPlayer, TCommandSource>,
    private val permissionService: PermissionService,
    private val pluginMessages: PluginMessages,
    private val registry: Registry,
    private val pluginInfo: PluginInfo
) {
    fun execute(context: CommandContext<TCommandSource>, playerClass: Class<*>): Int {
        if (!permissionService.hasPermission(context.source, registry.getOrDefault(CatalystKeys.TOGGLE_CHAT_PERMISSION))) {
            pluginMessages.noPermission.sendTo(context.source)
            return 0
        }
        if (!playerClass.isAssignableFrom(context.source!!::class.java)) {
            Component.text("Player only command!").sendTo(context.source)
            return 0
        }
        val player = context.source as TPlayer
        if (chatService.isDisabledForUser(player)) {
            chatService.toggleChatForUser(player)
            sendToggle(context.source, true)
            return 1
        }
        chatService.toggleChatForUser(player)
        sendToggle(context.source, false)
        return 1
    }

    private fun sendToggle(source: TCommandSource, enabled: Boolean) {
        val setting = if (enabled) "enabled" else "disabled"
        Component.text()
            .append(pluginInfo.prefix)
            .append(Component.text("Proxy-Wide chat has been ").color(NamedTextColor.GREEN))
            .append(Component.text(setting).color(NamedTextColor.GOLD))
            .append(Component.text("\nNote: This only affects you.").color(NamedTextColor.YELLOW))
            .sendTo(source)
    }
}
